genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -0,0 +1,363 @@
+    private final ExponentialBackOffTrigger trigger;
+    private final JobDirectoryManifestProtoConverter manifestProtoConverter;
+    private final StreamObserver<ServerControlMessage> responseObserver;
+    private final Semaphore concurrentTransfersSemaphore;
+    private final Set<FileTransfer> activeFileTransfers;
+
+    private StreamObserver<AgentManifestMessage> controlStreamObserver;
+    private String jobId;
+    private Path jobDirectoryPath;
+    private AtomicBoolean started = new AtomicBoolean();
+    private ScheduledFuture<?> scheduledTask;
+
+    GRpcAgentFileStreamServiceImpl(
+        final FileStreamServiceGrpc.FileStreamServiceStub fileStreamServiceStub,
+        final TaskScheduler taskScheduler,
+        final JobDirectoryManifestProtoConverter manifestProtoConverter
+    ) {
+        this.fileStreamServiceStub = fileStreamServiceStub;
+        this.taskScheduler = taskScheduler;
+        this.manifestProtoConverter = manifestProtoConverter;
+        this.trigger = new ExponentialBackOffTrigger(
+            ExponentialBackOffTrigger.DelayType.FROM_PREVIOUS_EXECUTION_BEGIN,
+            1000, //TODO make configurable
+            10000, //TODO make configurable
+            1.1f //TODO make configurable
+        );
+        this.responseObserver = new ServerControlStreamObserver(this);
+        this.concurrentTransfersSemaphore = new Semaphore(MAX_CONCURRENT_TRANSMIT_STREAMS);
+        this.activeFileTransfers = Sets.newConcurrentHashSet();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void start(final String claimedJobId, final Path jobDirectoryRoot) {
+        //Service can be started only once
+        if (!this.started.compareAndSet(false, true)) {
+            throw new IllegalStateException("Service can be started only once");
+        }
+        this.jobId = claimedJobId;
+        this.jobDirectoryPath = jobDirectoryRoot;
+        this.scheduledTask = this.taskScheduler.schedule(this::pushManifest, trigger);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void stop() {
+
+        if (this.started.compareAndSet(true, false)) {
+            this.scheduledTask.cancel(false);
+            this.scheduledTask = null;
+            this.discardCurrentStream(true);
+            while (!this.activeFileTransfers.isEmpty()) {
+                try {
+                    final FileTransfer fileTransfer = this.activeFileTransfers.iterator().next();
+                    if (this.activeFileTransfers.remove(fileTransfer)) {
+                        fileTransfer.completeTransfer(true, new InterruptedException("Shutting down"));
+                    }
+                } catch (NoSuchElementException | ConcurrentModificationException e) {
+                    // Swallow. Not unexpected, collection and state may change.
+                }
+            }
+        }
+    }
+
+    private synchronized void pushManifest() {
```

Comment:
You seem to have this method only called by the scheduled task. Why waste effort synchronizing it when only that thread will call this?

Reply:
It uses member variables that need synchronization with other thread.
For example `controlStreamObserver` is modified by `discardCurrentStream()`.